File: JavaScript/src/PayrollApplication.mjs
import PayCheck from './PayCheck.mjs'

export default class PayrollApplication {
  /**
   * @param {import('./Employee.mjs').default} employee
   * @param {number} workHours
   * @returns {import('./PayCheck.mjs').default}
   */
  payAmount(employee, workHours) {
    let result
    if (!employee.isSeparated()) {
      if (employee.isRetired()) {
        result = new PayCheck(0, "RET")
      } else {
        // Logic to compute amount
        let bonus = this.computeBonus(workHours)
        let regularAmount = this.computeRegularPayAmount(employee, workHours)
        let amount = bonus + regularAmount
        result = new PayCheck(amount, "EMP")
      }
    } else {
      result = new PayCheck(0, "SEP")
    }
    return result
  }

  /**
   * @private
   * @param {number} workHours
   * @returns {number}
   */
  computeBonus(workHours) {
    return workHours > 40 ? 1000 : 0
  }

  /**
   * @private
   * @param {import('./Employee.mjs').default} employee
   * @param {number} workHours
   * @returns {number}
   */
  computeRegularPayAmount(employee, workHours) {
    return employee.rate * workHours
  }
}

File: Python/payroll_application.py
from dataclasses import dataclass


@dataclass
class Employee:
    rate: int
    separated: bool
    retired: bool


@dataclass
class PayCheck:
    amount: float
    reasonCode: str


def _compute_regular_pay_amount(employee: Employee, work_hours: float) -> float:
    return employee.rate * work_hours


def _compute_bonus(work_hours: int) -> float:
    return 1000 if work_hours > 40 else 0


def pay_amount(employee: Employee, work_hours: int) -> PayCheck:
    result = None

    if not employee.separated:
        if employee.retired:
            result = PayCheck(0, "RET")
        else:
            # Logic to compute amount
            bonus = _compute_bonus(work_hours)
            regular_amount = _compute_regular_pay_amount(employee, work_hours)
            amount = bonus + regular_amount
            result = PayCheck(amount, "EMP")
    else:
        result = PayCheck(0, "SEP")

    return result
